### DISCUSSION TITLE
Consider providing a Promise version of the load function

### DISCUSSION THREAD

**Discussion by samcw** [katspaugh/wavesurfer.js#2780] [Ideas]
We used this library in our own project and had a very good experience, but one suggestion is to provide a promise version of the load function to enhance the flexibility of the library, we need to do something when audios loaded, although you can listen to ready events to achieve this goal, but lacks some flexibility. I would be happy to provide an mr to implement this functionality as well.

**Comment by katspaugh** [comment]
Sure, good idea.
Version 7 (alpha) already returns a promise btw.
It’s a bit tricky though, because there are several things going on when a file is loaded. The audio itself is loaded via a media element which doesn’t have a single point where you can say it’s completely loaded. There are events like loadedmetadata and canplay with slightly different semantics, but both would be good candidates.
And then we also load the file via xhr to decode it. Which has a pretty straight-forward completion point.

**Reply by katspaugh** [reply]
Thinking about this gave me the idea to set the audio src to the fetched data directly, so that there's just one fetch now (before it was two – one for rendering and one for playback).
The promise returned by the load function will now resolve when the audio is fully loaded and rendered.

**Reply by samcw** [reply]
Thank you for your reply.😀
